presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -80,7 +83,7 @@ public IterativeOptimizer(RuleStatsRecorder stats, StatsCalculator statsCalculat
     public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
     {
         // only disable new rules if we have legacy rules to fall back to
-        if (!SystemSessionProperties.isNewOptimizerEnabled(session) && !legacyRules.isEmpty()) {
+        if (legacyRulesEnabled.test(session) && !legacyRules.isEmpty()) {
```

Comment:
it does no longer be need to be legal for `legacyRules` to be empty when the test is true.

if i just want to have disabling a rule, i should use Rule#isEnabled for that)

Reply:
The goal is (and has always been) to be able to use this to migrate from legacy rules to new ones. When new rules are enabled, the legacy ones should be disabled (and vice-versa).